### PR TITLE
Prune Compound/Inter From Similar Blocks

### PR DESCRIPTION
**Pull request by chkngit**
## Description

 Use  valid previously encoded similar  blocks (block having the same location and shape , but different neighbours  as the current block). to prune the compounds/inter modes to search at MD

## Algorithm Details

1- If previous similar block is not compound, do not inject compound for current block
2- If previous similar block is not compound, do not inject compound for current block
else consider the compound modes up the similar’s one
3- If previous similar block is intra, do not inject inter current block

## Tests and performance
in context of enc-mode 0 using mixed mpeg/aom set (T00)

| Speed Deviation| BD-Rate Deviation|
| -- | --
| 2.52% | 0.00%|

### CODE REVIEW

File: Source/Lib/Encoder/Codec/EbProductCodingLoop.c [hunk 1]
```diff
@@ -5274,6 +5274,72 @@ void check_redundant_block(const BlockGeom *blk_geom, ModeDecisionContext *conte
...
+            {
+                *similar_blk_mds = blk_geom->similar_list.blk_mds_table[it];
+                *similar_blk_avail = 1;
+                break;
+            }
+        }
+    }
+}
+/******************************************************
+* Derive md  Settings  at he block level
```
Comment (andrey-norkin) [inline Source/Lib/Encoder/Codec/EbProductCodingLoop.c]: A typo

File: Source/Lib/Encoder/Codec/EbEncDecProcess.c [hunk 2]
```diff
@@ -1934,6 +1945,15 @@ EbErrorType signal_derivation_enc_dec_kernel_oq(SequenceControlSet * scs_ptr,
         else
             context_ptr->compound_types_to_try = MD_COMP_AVG;
     }
+#endif
+
+#if  INTRA_SIMILAR
```
Comment (andrey-norkin) [inline Source/Lib/Encoder/Codec/EbEncDecProcess.c]: We'll need to remove the macros after merging

**Comment by hassount** [comment]
@fouedb any issues on this PR?

File: Source/Lib/Encoder/Codec/EbEncDecProcess.c [hunk 3]
```diff
@@ -1921,6 +1921,17 @@ EbErrorType signal_derivation_enc_dec_kernel_oq(SequenceControlSet * scs_ptr,
                 ? FULL_PEL_REF_WINDOW_HEIGHT_EXTENDED
                 : FULL_PEL_REF_WINDOW_HEIGHT;
     }
+#if COMP_SIMILAR
+    //comp_similar_mode
+    //0: OFF
+    //1: If previous similar block is not compound, do not inject compound
+    //2: If previous similar block is not compound, do not inject compound
+    //   else consider the compound modes up the similar�s one
```
Comment (fouedb) [inline Source/Lib/Encoder/Codec/EbEncDecProcess.c]: Consider instead "else consider the compound modes up the mode for the similar block"

File: Source/Lib/Encoder/Codec/EbProductCodingLoop.c [hunk 4]
```diff
@@ -5274,6 +5274,72 @@ void check_redundant_block(const BlockGeom *blk_geom, ModeDecisionContext *conte
     }
 }
 
+#if COMP_SIMILAR
+/*
+   search for a valid previously encoded similar
+   block (block having the same location and shape , but different neighbours
```
Comment (fouedb) [inline Source/Lib/Encoder/Codec/EbProductCodingLoop.c]: Consider instead: 
   "block (block having the same location and shape as the current block, but where neighboring blocks are different from those for the current block)"

File: Source/Lib/Encoder/Codec/EbProductCodingLoop.c [hunk 5]
```diff
@@ -5274,6 +5274,72 @@ void check_redundant_block(const BlockGeom *blk_geom, ModeDecisionContext *conte
...
+                *similar_blk_avail = 1;
+                break;
+            }
+        }
+    }
+}
+/******************************************************
+* Derive md  Settings  at he block level
+******************************************************/
+EbErrorType signal_derivation_block(
```
Comment (fouedb) [inline Source/Lib/Encoder/Codec/EbProductCodingLoop.c]: Add a description of the function.

File: Source/Lib/Encoder/Codec/EbProductCodingLoop.c (hunk 1 above)
Comment (chkngit) [reply Source/Lib/Encoder/Codec/EbProductCodingLoop.c]: done

File: Source/Lib/Encoder/Codec/EbEncDecProcess.c (hunk 3 above)
Comment (chkngit) [reply Source/Lib/Encoder/Codec/EbEncDecProcess.c]: done

File: Source/Lib/Encoder/Codec/EbProductCodingLoop.c (hunk 4 above)
Comment (chkngit) [reply Source/Lib/Encoder/Codec/EbProductCodingLoop.c]: done

**Pull request merged**